File: refactor.js
const fs = require('fs');
const path = require('path');

const appJsPath = 'app.js';
const adapterJsPath = path.join('src', 'services', 'dataAdapter.js');

let content = fs.readFileSync(appJsPath, 'utf8');

function extractFunction(name, source) {
    const regex1 = new RegExp(`(?:async\s+)?function\s+${name}\s*\([^)]*\)\s*\{`, 'm');
    const regex2 = new RegExp(`(?:const|let|var|window\.)\s*${name}\s*=\s*(?:async\s+)?(?:function)?\s*\([^)]*\)\s*(?:=>\s*)?\{`, 'm');
    
    let match = source.match(regex1) || source.match(regex2);
    if (!match) return { body: null, newSource: source };
    
    const startIdx = match.index;
    let braceCount = 0;
    let inString = false;
    let stringChar = '';
    let escape = false;
    
    for (let i = startIdx + match[0].length - 1; i < source.length; i++) {
        const char = source[i];
        
        if (escape) {
            escape = false;
            continue;
        }
        
        if (char === '') {
            escape = true;
            continue;
        }
        
        if (inString) {
            if (char === stringChar) {
                inString = false;
            }
        } else {
            if (char === "'" || char === '"' || char === '`') {
                inString = true;
                stringChar = char;
            } else if (char === '{') {
                braceCount++;
            } else if (char === '}') {
                braceCount--;
                if (braceCount === 0) {
                    const endIdx = i + 1;
                    const body = source.substring(startIdx, endIdx);
                    const newSource = source.substring(0, startIdx) + source.substring(endIdx);
                    return { body, newSource };
                }
            }
        }
    }
    return { body: null, newSource: source };
}

const funcsToMove = [
    'loadAsanaLibrary', 
    'normalizeAsana', 
    'normalizeAsanaRow', 
    'normalizePlate', 
    'parsePlates', 
    'normaliseAsanaId'
];

let extractedFuncs = [];
let foundFuncNames = [];

for (const func of funcsToMove) {
    const { body, newSource } = extractFunction(func, content);
    if (body) {
        extractedFuncs.push(body);
        foundFuncNames.push(func);
        content = newSource;
    }
}

let adapterCode = `import { supabase } from './supabaseClient.js';
import { parseHoldTimes } from '../utils/parsing.js';

${extractedFuncs.join('

')}

export { ${foundFuncNames.join(', ')} };
`;

fs.writeFileSync(adapterJsPath, adapterCode, 'utf8');

const importStatement = `import { ${foundFuncNames.join(', ')} } from './src/services/dataAdapter.js';
`;
const importInsertPos = content.indexOf('import { supabase } from');
if (importInsertPos !== -1) {
    content = content.substring(0, importInsertPos) + importStatement + content.substring(importInsertPos);
} else {
    content = importStatement + content;
}

fs.writeFileSync(appJsPath, content, 'utf8');

console.log(`Successfully moved ${foundFuncNames.length} functions to ${adapterJsPath}`);

File: refactor.py
import re
import os

app_js_path = 'app.js'
adapter_js_path = os.path.join('src', 'services', 'dataAdapter.js')

with open(app_js_path, 'r', encoding='utf-8') as f:
    content = f.read()

def extract_function(name, content):
    pattern = re.compile(rf'(?:async\s+)?function\s+{name}\s*\([^)]*\)\s*{{', re.MULTILINE)
    match = pattern.search(content)
    if not match:
        pattern = re.compile(rf'(?:const|let|var|window\.)\s*{name}\s*=\s*(?:async\s+)?(?:function)?\s*\([^)]*\)\s*(?:=>\s*)?{{', re.MULTILINE)
        match = pattern.search(content)
    
    if not match:
        return None, content
        
    start_idx = match.start()
    brace_count = 0
    in_string = False
    string_char = ''
    escape = False
    
    for i in range(match.end() - 1, len(content)):
        char = content[i]
        
        if escape:
            escape = False
            continue
            
        if char == '':
            escape = True
            continue
            
        if in_string:
            if char == string_char:
                in_string = False
        else:
            if char in ["'", '"', '`']:
                in_string = True
                string_char = char
            elif char == '{':
                brace_count += 1
            elif char == '}':
                brace_count -= 1
                if brace_count == 0:
                    end_idx = i + 1
                    func_body = content[start_idx:end_idx]
                    new_content = content[:start_idx] + content[end_idx:]
                    return func_body, new_content
                    
    return None, content

funcs_to_move = [
    'loadAsanaLibrary', 
    'normalizeAsana', 
    'normalizeAsanaRow', 
    'normalizePlate', 
    'parsePlates', 
    'normaliseAsanaId'
]

extracted_funcs = []
for func in funcs_to_move:
    body, content = extract_function(func, content)
    if body:
        extracted_funcs.append(body)

# Write to dataAdapter.js
adapter_code = """import { supabase } from './supabaseClient.js';
import { parseHoldTimes } from '../utils/parsing.js';

""" + "

".join(extracted_funcs) + "

"

# Export the functions
exports = ", ".join([f for f in funcs_to_move if any(f in body for body in extracted_funcs)])
adapter_code += f"export {{ {exports} }};
"

with open(adapter_js_path, 'w', encoding='utf-8') as f:
    f.write(adapter_code)

# Update app.js
import_statement = f"import {{ {exports} }} from './src/services/dataAdapter.js';
"
# Find a good place to insert the import in app.js
import_insert_pos = content.find("import { supabase } from")
if import_insert_pos != -1:
    content = content[:import_insert_pos] + import_statement + content[import_insert_pos:]
else:
    content = import_statement + content

with open(app_js_path, 'w', encoding='utf-8') as f:
    f.write(content)

print(f"Successfully moved {len(extracted_funcs)} functions to {adapter_js_path}")
